### PR TITLE
Fixes #16737: rl variables throw error if variable names are the same

### PR DESCRIPTION
**Pull request by gpoblon**
https://issues.rudder.io/issues/16737

### CODE REVIEW

**Comment by gpoblon** [comment]
rebased + got it ready to be merged into master

File: rudder-lang/src/parser.rs [hunk 1]
```diff
@@ -599,9 +600,24 @@ fn pstruct(i: PInput) -> PResult<HashMap<String, PValue>> {
     )(i)
 }
 
+/// Alternative version of pstruct based on "." rather than braces.
+/// Used for the agent
+/// A struct is stored in a HashMap
+// fn pstruct_agent(i: PInput) -> PResult<HashMap<String, PValue>> {
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: Why is this here ?

File: rudder-lang/src/parser.rs [hunk 2]
```diff
@@ -766,6 +785,41 @@ fn pvariable_definition(i: PInput) -> PResult<(Token, PValue)> {
     )(i)
 }
 
+fn fill_map_rec<'src>(mut tokens: std::iter::Peekable<std::slice::Iter<Token<'src>>>) -> HashMap<String, PValue<'src>> {
+    let mut map: HashMap<String, PValue> = HashMap::new();
+    if let Some(tk) = tokens.next() {
+        let tk_str = tk.fragment().to_owned();
+        if tokens.peek().is_some() {
+            map.insert(tk_str, PValue::Struct(fill_map_rec(tokens)));
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: unusual way to iterate over an iterator

File: rudder-lang/src/parser.rs (hunk 1 above)
Comment (gpoblon) [reply rudder-lang/src/parser.rs]: It should not anymore

**Comment by gpoblon** [comment]
Commit modified

**Comment by gpoblon** [comment]
PR updated with a new commit
switched recursive to fold

File: rudder-lang/src/parser.rs (hunk 2 above)
Comment (gpoblon) [reply rudder-lang/src/parser.rs]: switched to a rev().fold()

**Comment by gpoblon** [comment]
PR rebased

File: rudder-lang/src/parser.rs [hunk 3]
```diff
@@ -766,6 +771,35 @@ fn pvariable_definition(i: PInput) -> PResult<(Token, PValue)> {
     )(i)
 }
 
+fn pvalue_varagent(i: PInput) -> PResult<PValue> {
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: This parser needs a doc comment too

**Comment by gpoblon** [comment]
PR updated with a new commit
parser comment

File: rudder-lang/src/parser.rs (hunk 3 above)
Comment (gpoblon) [reply rudder-lang/src/parser.rs]: yes, done

**Comment by peckpeck** [comment]
@gpoblon  conflicts -> you ust rebase

**Comment by peckpeck** [comment]
OK, squash merging this PR

**Comment by peckpeck** [comment]
Merged, but due to multiple rebases this might not have been detected
